pkg/apis/pipeline/v1beta1/task_validation.go

```diff
@@ -286,102 +250,73 @@ func ValidateResourcesVariables(steps []Step, resources *TaskResources) *apis.Fi
-				return err
-			}
-		}
-		for i, arg := range step.Args {
-			if err := validateTaskVariable(fmt.Sprintf("arg[%d]", i), arg, prefix, vars); err != nil {
-				return err
-			}
-		}
-		for _, env := range step.Env {
-			if err := validateTaskVariable(fmt.Sprintf("env[%s]", env.Name), env.Value, prefix, vars); err != nil {
-				return err
-			}
-		}
-		for i, v := range step.VolumeMounts {
-			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].Name", i), v.Name, prefix, vars); err != nil {
-				return err
-			}
-			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].MountPath", i), v.MountPath, prefix, vars); err != nil {
-				return err
-			}
-			if err := validateTaskVariable(fmt.Sprintf("volumeMount[%d].SubPath", i), v.SubPath, prefix, vars); err != nil {
-				return err
-			}
-		}
+func validateStepArrayUsage(step Step, prefix string, vars sets.String) *apis.FieldError {
+	errs := validateTaskNoArrayReferenced(step.Name, prefix, vars).ViaField("name")
+	errs = errs.Also(validateTaskNoArrayReferenced(step.Image, prefix, vars).ViaField("image"))
+	errs = errs.Also(validateTaskNoArrayReferenced(step.WorkingDir, prefix, vars).ViaField("workingDir"))
+	errs = errs.Also(validateTaskNoArrayReferenced(step.Script, prefix, vars).ViaField("script"))
+	for i, cmd := range step.Command {
+		errs = errs.Also(validateTaskArraysIsolated(cmd, prefix, vars).ViaFieldIndex("command", i))
 	}
-	return nil
+	for i, arg := range step.Args {
+		errs = errs.Also(validateTaskArraysIsolated(arg, prefix, vars).ViaFieldIndex("args", i))
+
+	}
+	for _, env := range step.Env {
+		errs = errs.Also(validateTaskNoArrayReferenced(env.Value, prefix, vars).ViaFieldKey("env", env.Name))
+	}
+	for i, v := range step.VolumeMounts {
+		errs = errs.Also(validateTaskNoArrayReferenced(v.Name, prefix, vars).ViaField("name").ViaFieldIndex("volumeMount", i))
+		errs = errs.Also(validateTaskNoArrayReferenced(v.MountPath, prefix, vars).ViaField("mountPath").ViaFieldIndex("volumeMount", i))
+		errs = errs.Also(validateTaskNoArrayReferenced(v.SubPath, prefix, vars).ViaField("subPath").ViaFieldIndex("volumeMount", i))
+	}
+	return errs
+}
+
+func validateVariables(steps []Step, prefix string, vars sets.String) (errs *apis.FieldError) {
+	for idx, step := range steps {
+		errs = errs.Also(validateStepVariables(step, prefix, vars).ViaFieldIndex("steps", idx))
+	}
+	return errs
+}
+
+func validateStepVariables(step Step, prefix string, vars sets.String) *apis.FieldError {
+	errs := validateTaskVariable(step.Name, prefix, vars).ViaField("name")
+	errs = errs.Also(validateTaskVariable(step.Image, prefix, vars).ViaField("image"))
+	errs = errs.Also(validateTaskVariable(step.WorkingDir, prefix, vars).ViaField("workingDir"))
+	errs = errs.Also(validateTaskVariable(step.Script, prefix, vars).ViaField("script"))
+	for i, cmd := range step.Command {
+		errs = errs.Also(validateTaskVariable(cmd, prefix, vars).ViaFieldIndex("command", i))
```

Comment:
NIT: could have been into a common function (line 296 - 308) which is also appearing above